frozen::map of std::variant compilation error

when trying to have a map of variant with std::pair or frozen::set in the variant, compilation fail

Compiler : gcc 9.2.1 with -std=c++2a


`
#include <utility>
#include  <variant>
#include <frozen/map.h>
#include <frozen/string.h>
#include <frozen/set.h>

using test_variant_t = std::variant<float , frozen::set<int, 2> >;

constexpr auto conf_dict =
  frozen::make_map<frozen::string, test_variant_t> ({
					{"A", 10.0f}
		     });

int main (void) {}


`

compilation fail with : 
 _
frozen/include/frozen/bits/algorithms.h:89:5: error: use of deleted function 'std::variant<_Types>& std::variant<_Types>::operator=(const std::variant<_Types>&) [with _Types = {float, frozen::set<int, 2, std::less<int> >}]'
   89 |   a = b;
_

Unfortunately, it's not.

I am afraid that the problem is in std::variant, and we do not have frozen::variant.

error message is :
`In file included from /home/alex/DEV/STM32/frozen/include/frozen/map.h:28,
                 from /tmp/testFrozen.cpp:3:
/home/alex/DEV/STM32/frozen/include/frozen/bits/algorithms.h: In instantiation of ‘constexpr void frozen::bits::cswap(T&, T&) [with T = std::variant<float, frozen::set<int, 2, std::less<int> > >]’:
/home/alex/DEV/STM32/frozen/include/frozen/bits/algorithms.h:96:8:   required from ‘constexpr void frozen::bits::cswap(std::pair<_T1, _T2>&, std::pair<_T1, _T2>&) [with T = frozen::string; U = std::variant<float, frozen::set<int, 2, std::less<int> > >]’
/home/alex/DEV/STM32/frozen/include/frozen/bits/algorithms.h:114:8:   required from ‘constexpr Iterator frozen::bits::partition(Iterator, Iterator, const Compare&) [with Iterator = std::pair<frozen::string, std::variant<float, frozen::set<int, 2, std::less<int> > > >*; Compare = frozen::impl::CompareKey<std::less<frozen::string> >]’
/home/alex/DEV/STM32/frozen/include/frozen/bits/algorithms.h:128:37:   required from ‘constexpr void frozen::bits::quicksort(Iterator, Iterator, const Compare&) [with Iterator = std::pair<frozen::string, std::variant<float, frozen::set<int, 2, std::less<int> > > >*; Compare = frozen::impl::CompareKey<std::less<frozen::string> >]’
/home/alex/DEV/STM32/frozen/include/frozen/bits/algorithms.h:138:12:   required from here
/home/alex/DEV/STM32/frozen/include/frozen/bits/algorithms.h:89:5: error: use of deleted function ‘std::variant<_Types>& std::variant<_Types>::operator=(const std::variant<_Types>&) [with _Types = {float, frozen::set<int, 2, std::less<int> >}]’
   a = b;
   ~~^~~
In file included from /tmp/testFrozen.cpp:2:
/usr/include/c++/8/variant:1083:16: note: ‘std::variant<_Types>& std::variant<_Types>::operator=(const std::variant<_Types>&) [with _Types = {float, frozen::set<int, 2, std::less<int> >}]’ is implicitly deleted because the default definition would be ill-formed:
       variant& operator=(const variant&) = default;
                ^~~~~~~~
/usr/include/c++/8/variant:1083:16: error: use of deleted function ‘std::_Enable_copy_move<true, false, true, false, _Tag>& std::_Enable_copy_move<true, false, true, false, _Tag>::operator=(const std::_Enable_copy_move<true, false, true, false, _Tag>&) [with _Tag = std::variant<float, frozen::set<int, 2, std::less<int> > >]’
In file included from /usr/include/c++/8/variant:38,
                 from /tmp/testFrozen.cpp:2:
/usr/include/c++/8/bits/enable_special_members.h:244:5: note: declared here
     operator=(_Enable_copy_move const&) noexcept                    = delete;
     ^~~~~~~~
In file included from /home/alex/DEV/STM32/frozen/include/frozen/map.h:28,
                 from /tmp/testFrozen.cpp:3:
/home/alex/DEV/STM32/frozen/include/frozen/bits/algorithms.h:90:5: error: use of deleted function ‘std::variant<_Types>& std::variant<_Types>::operator=(const std::variant<_Types>&) [with _Types = {float, frozen::set<int, 2, std::less<int> >}]’
   b = tmp;
   ~~^~~~~
`

--- a/include/frozen/bits/algorithms.h
+++ b/include/frozen/bits/algorithms.h
@@ -197,6 +197,30 @@ constexpr bool binary_search(ForwardIt first, const T &value,
   return (!(where == first + N) && !(compare(value, *where)));
 }
 
+
+template<class InputIt1, class InputIt2>
+constexpr bool equal(InputIt1 first1, InputIt1 last1, InputIt2 first2)
+{
+  for (; first1 != last1; ++first1, ++first2) {
+    if (!(*first1 == *first2)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template<class InputIt1, class InputIt2>
+constexpr bool lexicographical_compare(InputIt1 first1, InputIt1 last1, InputIt2 first2, InputIt2 last2)
+{
+  for (; (first1 != last1) && (first2 != last2); ++first1, ++first2) {
+    if (*first1 < *first2)
+      return true;
+    if (*first2 < *first1)
+      return false;
+  }
+  return (first1 == last1) && (first2 != last2);
+}
+
 } // namespace bits
 } // namespace frozen
 

--- a/include/frozen/set.h
+++ b/include/frozen/set.h
@@ -32,8 +32,8 @@ namespace frozen {
 
 template <class Key, std::size_t N, class Compare = std::less<Key>> class set {
   using container_type = bits::carray<Key, N>;
-  Compare const less_than_;
-  container_type const keys_;
+  Compare less_than_;
+  container_type keys_;
 
 public:
   /* container typedefs*/
@@ -128,11 +128,19 @@ public:
   constexpr const_reverse_iterator crbegin() const { return keys_.crbegin(); }
   constexpr const_reverse_iterator rend() const { return keys_.rend(); }
   constexpr const_reverse_iterator crend() const { return keys_.crend(); }
+
+  /* comparison */
+  constexpr bool operator==(set const& rhs) const { return bits::equal(begin(), end(), rhs.begin()); }
+  constexpr bool operator!=(set const& rhs) const { return !(*this == rhs); }
+  constexpr bool operator<(set const& rhs) const { return bits::lexicographical_compare(begin(), end(), rhs.begin(), rhs.end()); }
+  constexpr bool operator<=(set const& rhs) const { return (*this < rhs) || (*this == rhs); }
+  constexpr bool operator>(set const& rhs) const { return bits::lexicographical_compare(rhs.begin(), rhs.end(), begin(), end()); }
+  constexpr bool operator>=(set const& rhs) const { return (*this > rhs) || (*this == rhs); }
 };
 
 template <class Key, class Compare> class set<Key, 0, Compare> {
   using container_type = bits::carray<Key, 0>; // just for the type definitions
-  Compare const less_than_;
+  Compare less_than_;
 
 public:
   /* container typedefs*/

--- a/tests/test_set.cpp
+++ b/tests/test_set.cpp
@@ -264,3 +264,14 @@ TEST_CASE("frozen::set constexpr", "[set]") {
   static_assert(!ce.count(0), "");
   static_assert(ce.find(0) == ce.end(), "");
 }
+
+TEST_CASE("frozen::set of frozen::set", "[set]") {
+  using s1 = frozen::set<unsigned, 1>;
+  constexpr frozen::set<s1, 2> ce = {{3}, {11}};
+  static_assert(*ce.begin() == s1({3}), "");
+  static_assert(*(ce.begin() + 1) == s1({11}), "");
+  static_assert(ce.size() == 2, "");
+  static_assert(ce.count(s1({3})), "");
+  static_assert(!ce.count(s1({0})), "");
+  static_assert(ce.find(s1({0})) == ce.end(), "");
+}

